### ISSUE TITLE
Revisiting behavior of dials

### ISSUE THREAD

**Issue by tasota** [BirdBrainTechnologies/flutter-kindle-sttr2#43]
Emily wrote up some thoughts from playing with the new version of the app with LED controls implemented (see #39 #40 #42) which I think warrants a general discussion on how we want the dials to behave. I figure that since the discussion is all interconnected we could consolidate our discussions to this issue for now.

For remainders, dial 2 controls the divisor (columns) and its range of values is always 1-12. For dial 1, the dividend's range is from 0 to 12 * the divisor since this is the largest number that can be displayed on the grid.

For subtraction, dial 1 controls the first term and has values from 0 to 120. Dial 2 controls the second term and has values from 0 to the first term (with inverted values, so turning dial 2 counter-clockwise increases this term).

**Comment by bambibrewer** [comment]
I think it makes sense to combine the discussions here, as I agree they are related. I think the easiest is #40 - it was suggested in the focus group that the dial increase the number subtracted as you turn it counterclockwise because that increases the total. However, if people feel that is not working, then I think it is pretty straightforward to just increase that number as you turn the dial clockwise.

I think #39 is trickier - the left number can't be less than the right one, so at some point you will have to adjust the right number even though you are turning the left dial (because we can't have negative numbers). One option would be to keep the right number the same until the left number is equal to it (7-7 = 0, for example). At that point, if you keep decreasing the left number, both of them decrease. Do people think that would make more sense? There will still be some weirdness - when you start to increase the left dial again, the LEDs will appear as the "subtracted" color. 

This does also emphasize to me that we want to recommend a specific sequence in the instructions/professional development - use the left dial to set the first number, then use the right dial to set the second. Take them both back to 0 at the end of the problem before starting the next one.

**Comment by bambibrewer** [comment]
For division, I really can't picture any other way of doing it. I totally see the point, but one dial has to set the divisor, and then other dial has to be scaled based on the divisor. For example, if the divisor is 2, the dividend (number we are dividing) can't be 144 because it won't fit on the array.

**Comment by tasota** [comment]
Yeah definitely in agreement with #40 since this is the only instance of increasing from turning counter-clockwise. I can resolve that one rather quickly if we're all in agreement with that one.

For #39 and #42 an alternative that I suggested to Emily was that we do not scale the other value until the other dial position changes. However, as Bambi describes above, at a certain point of turning the dial for subtraction this is unavoidable (since you cannot have a negative result). This forced change is only when values decrease for subtraction and remainders (to avoid inability to represent). When you increase values you should be safe (for example, going from divisor 2 to 3 can retain 24, since 24 / 2 = 12 R 0 and 24 / 3 = 8 R 1, but we cannot represent 24 / 1 = 24 R 0)

**Comment by ehamner** [comment]
Bambi: "it was suggested in the focus group that the dial increase the number subtracted as you turn it counterclockwise because that increases the total. However, if people feel that is not working, then I think it is pretty straightforward to just increase that number as you turn the dial clockwise."

Mike: "Yeah definitely in agreement since this is the only instance of increasing from turning counter-clockwise. I can resolve that one rather quickly if we're all in agreement with that one."

**Comment by ehamner** [comment]
I am trying to explore other options (which may or may not be good). Here is one idea for #39. 

Keep the dials and display values in sync. Make each dial control only one number. Increasing as the dial turns clockwise.

When this results in an equation that would be a negative number, such as 100-150:
-display 100 faded circles on the LED array
-write the equation as 100 - 150 = less than 0
-write the 100 in the equation as it normally looks
-write the 150 and "less than 0" visually different so that you would see you are creating an equation that doesn't "work." I imagine them as greyed out boxes with greyed out text.

**Comment by ehamner** [comment]
Brainstorming for #42 

Left dial controls dividend (rows). Right dial controls divisor (columns). App displays the correct equation even if it can not be represented on the array. The array visually indicates that the equation can not be displayed. Possible ideas for displaying an overflowing equation:
1) the LEDs at the end of the rows or columns that would extend off the grid blink
2) LEDs "scroll", indicated by different colors every 12 LEDs
3) Display a message such as Full on the array (fits in 11 across, 5 high)
4) blink the last 3 LEDs in the row/column in a pattern
5) combination of some of the above. Dots indicating the problem light up one color "behind" the word Full or an arrow in another color. End LEDs blink.
6) display an arrow on the LED array along with the visible portion of the equation

**Comment by ehamner** [comment]
Tested with Katie:

Overall she liked it how it was and thought it was a lot of fun. We went through all the explore modes. She moved the dials very quickly and liked seeing the lights change. She wasn't bothered by the things I was bothered by because she moved the dials so quickly back and forth. After watching her use it, I feel ready to pilot the explore modes as they are.

Other notes:
- When the LED array turned on and all the LEDS were on in different colors, she thought it's really cool!

- Subtraction/explore: 
"You should make instructions: First make the number. Then make the number you are taking away."
"Ok that is sort of helpful because I was wondering [what to take away from 120] to equal the same amount."

- Division/explore: 
Would like to have more colors.
"I know it's going to be 4 because it's 4 rows."

- Addition/explore:
"Oh yay colorful."
The last two rows on the app don't light up and she wanted it to match the array. (I believe that's already an issue.)
Likes the overlay.
"My only complaint is it's really hard when you only have one number left. It's hard to get the exact number you want."

- Division with Remainders/explore
This was a new concept for her. I did some explaining. She was able to explain to me that the remainder was "the dots left over." How are they left over? "They don't have a group to go into. It's not even."
Personal preference, she liked the look of the blue LEDs more than the pink.

- Divide 24 evenly in 5 ways:
"It would be helpful to have [an overlay]. Where is 24?" Eventually worked out 4 x 6. But she knew 4 x 6 would be 24 already. Nothing wrong with this mode, but it made me think of a mode between explore and make (is this make?), where you set 24 or the app light up 24, and then you played with the dials to make the rows and columns, but the number of dots stayed fixed.

- Also tried Find and liked that. This one got the most math focus out of her. Create was fun too. If more parts of the app had been ready, I think she would have continued to play with it for a while. I'll be curious to see how long she's willing to practice on it once the problem sets are in place.

**Comment by bambibrewer** [comment]
I have a suggestion about dials for addition. Right now, the left knob sets the first number and then the right know sets the second number from 0 - (120 - number 1). This avoids overflowing the array, but it means that turning the first knob also changes the second number, unless it is 0. That is a little weird, and I think we can make things a little more intuitive in a couple of ways:

1. In levels 1 and 2, numbers are restricted to 0-10 and 0-50, respectively. For these levels, there is no danger of overflow, so each knob can have a constant mapping.
2. In level 3, the left knob can be scaled constantly to 0-100. I would suggest that the right knob have a scaling from 20-100, but if the sum is > 120, then the app just caps the value of the second number so that the sum doesn't overflow the grid. This still results in some situations where adjusting the left knob changes both numbers, but this will only happen when the sum is 120, whereas now changing the first number changes the second one even if  you are not overflowing the grid.

What do you think, Mike? If it is helpful, this is something I could try out in the code (just let me know - I don't want to mess with your code base if it isn't something that would be useful to you).

**Comment by ehamner** [comment]
I like Bambi's suggestion for addition.

**Comment by tasota** [comment]
Sorry, just got a chance to catch up with all of this!

So, potentially restrict the range of the dials based on the current level? Will this just be for addition or **possibly for the other modes as well?**

And I think I like the suggestion for Level 3 as well. It sounds complicated when written out but when I imagine that interaction it makes sense to me.

**Comment by bambibrewer** [comment]
@tasota To try to clarify this (and other things), I took a stab at updating the (renamed) GlowGrid Details document to include dial settings for each mode and level: https://docs.google.com/document/d/11apdt4sJ6Xdra-IQ6dY2wHAa1cTLTNXtNEhKIYu7-RE/edit?usp=sharing

Emily and Lauren, if you could take a look and see what you think, that would be great. Mike, I would love to talk through it with you if you have time this week to see if you think things make sense, are technically feasible, if I left anything out, etc. Maybe Friday before the Flutter meeting?

**Comment by tasota** [comment]
Thank you Bambi. And yes I'm free all morning up until the flutter meeting on Friday so we can discuss whenever that morning (I should be in by 8:30)

**Comment by ehamner** [comment]
Still thinking about Division with Remainders, but I like everything else how Bambi has written it out in the GlowGrid Details document.

**Comment by ehamner** [comment]
I've added another option to consider for Division with Remainders.

Dial Settings: Right dial scaled 1-12, sets the number you are dividing by. Left dial scaled 1-144. While Left / Right > 12, then Right = Right + 1.

I think that's maybe more similar to the way we are proposing to do the other arithmetic operators. The scale for both dials is fixed, but sometimes the displayed value of the right dial has to be adjusted. (Maybe there's a better equation to use than the while loop format I was using?)

**Comment by tasota** [comment]
@ehamner Alright, I think I see what you are saying here. If I understand correctly, we will basically be changing the lower bound of the right dial's range as the left dial's value increases.

For example, if left is like 37, then the right dial would take on values from 4 to 12. Does this mean that if Right dial was in position less than 4, it would just round up to 4?

Also @ehamner will you be joining our discussions on Friday morning?

**Comment by ehamner** [comment]
Yes on all accounts.

Can you add me to the calendar event for Friday?

**Comment by tasota** [comment]
Ok, Added!

I had it from 9:30-11 on my calendar mostly to remind myself to be there by 9:30. Not sure if we will actually take that long to discuss.

**Comment by bambibrewer** [comment]
Thanks for adding that, Mike! I will be there.

**Comment by tasota** [comment]
See GlowGrid Details document for dial explanation. Basically, we agreed to maintain a constant range for the dials and round the right dial value if it goes over the maximum. For remainders, the smallest value that the divisor can take on is the smallest value that can be displayed on the grid.

**Comment by bambibrewer** [comment]
Skye and I were just playing with division with remainders, and things still seem a little strange. I think the main issue is that when you start with both dials all the way to the left (which people will do, if they have used the other sections) and then you start changing the left dial, both the rows and the columns change. I think it is made slightly stranger by the fact that you can't set the right dial first (because all the lights start off).
What if the right dial always set the number of columns from 1-12, and then the left dial selects the total number from 1-144, but if left dial > number of columns*12, the left dial is reset to number of columns*12? It seems like it might make what happens a little more predictable.

**Comment by tasota** [comment]
For the instruction, hasn't it been the case that the left dial is adjusted first, followed by the right dial? I would imagine you want the most control in fixing your dividend for the equation.

**Comment by bambibrewer** [comment]
That is a good point, Mike. The other sections do have people set the left dial first. In both division sections, that is causing me a little confusion in working on the manual, because you can't really set the dividend first. You have to set the divisor first before it makes sense to set the dividend. Does it make sense to have the left dial control the divisor and then the right dial controls the dividend?

**Comment by bambibrewer** [comment]
Maybe we should make the left dial the divisor for both division modes and see how that changes Division with Remainders? That section can wait a couple of weeks anyway, given that it won't be used in the second and third grade pilots.

**Comment by tasota** [comment]
@tasota to switch divisor and quotient generated in Division problem (this will make left dial control divisor, right dial control quotient)

Example: right now 63 / 9 = 7 <=>  7 * 9 = 63;; switch to be 63 / 7 = 9 <=> 7 * 9 / 63

**Comment by bambibrewer** [comment]
@tasota Were any changes ever made to Division with Remainders? Should we not use that section at PD tonight? Right not, it seems like the left dial is not changing the divisor in Division with Remainders, and I am wondering whether I am just out of date.

**Comment by bambibrewer** [comment]
I think the display looks like it should, but the left dial was supposed to control the divisor and the right dial the number being divided. Right now they are swapped, I think.

**Comment by tasota** [comment]
Hey, yes I think you are out of date. I was just looking at this issue and was considering closing it. I've implemented it now so the divisor is the number of rows instead of number of columns. I believe this was what we wanted?

`pd-3` branch should be what will be used tonight but we can confirm this later when we meet beforehand.

**Comment by bambibrewer** [comment]
Okay, I just updated, and I think the issue is that the left dial controls the divisor in Division, and I think that should also be the case in Division with Remainders. Everything works well, but I think the left and right dials should be switched. Is that possible before this afternoon?

**Comment by tasota** [comment]
@bambibrewer I'll work on this now, should be ready by the time we meet

**Comment by bambibrewer** [comment]
@tasota Can we also have the minimum number of rows/columns be 1, like in multiplication/division? Just to make things a little more consistent?

**Comment by tasota** [comment]
Yes I will also take this into consideration while updating this. Thank you

**Comment by bambibrewer** [comment]
@tasota Here are the changes we talked about! Things are working well, I think we just want to make a couple of minor changes to make things a little more intuitive. The left dial should scale from 1-12, just like it does now. The right dial should go from 1-144, with adjustments for two situations:

1. If the value of the right dial is less than the value of the left dial, set right = left (so you can always see the number of rows to help people select the divisor first).

2. If the value of the right dial is > 12*left, set right = 12*left. Otherwise, changing the right dial also changes the number of rows, which is confusing.

**Comment by tasota** [comment]
I was just playing with this implementation (commit 887df0357d6c9612626c51799006fd41eedba068) and it feels a lot nicer now, I think.

The master branch is now updated with this if you want to give it a try!

**Comment by bambibrewer** [comment]
Nevermind, I now see that it is the master branch!
[image: BirdBrain Technologies] <http://birdbraintechnologies.com/>

*Bambi Brewer, Ph.D.*
Director of Engineering, BirdBrain Technologies

888-371-6161 <1-888-371-6161> | bambi@birdbraintechnologies.com
www.birdbraintechnologies.com <http://birdbraintechnologies.com/>
544 Miltenberger St, Pittsburgh, PA 15219

[image: Facebook] <http://www.facebook.com/birdbraintech> [image: Instagram]
<http://instagram.com/birdbraintech/> [image: Twitter]
<http://twitter.com/birdbraintech> [image: YouTube]
<http://www.youtube.com/channel/UCxjgGxBG2QhymwC2FHpt3zw>  *@birdbraintech*
